audio-echo/app/src/main/java/com/google/sample/echo/MainActivity.java

```diff
@@ -120,11 +133,46 @@ private void queryNativeAudioParameters() {
         nativeSampleFormat ="";    //TODO: find a way to get the native audio format
     }
     private void updateNativeAudioUI() {
-        status_view.setText("nativeSampleRate    = " + nativeSampleRate + "\n" +
-                "nativeSampleBufSize = " + nativeSampleBufSize + "\n" +
-                "nativeSampleFormat  = " + nativeSampleFormat);
+        if(permissionGranted) {
+            status_view.setText("nativeSampleRate    = " + nativeSampleRate + "\n" +
+                    "nativeSampleBufSize = " + nativeSampleBufSize + "\n" +
+                    "nativeSampleFormat  = " + nativeSampleFormat);
+        } else {
+            status_view.setText("Error: Resource Permission not Granted" +
+                                "could not record Audio\n"    +
+                                 "Please restart App and grant the necessary permissions");
+        }
+    }
+    private void requestResourcePermission() {
+        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.RECORD_AUDIO) !=
+                PackageManager.PERMISSION_GRANTED) {
+            ActivityCompat.requestPermissions(this, new String[]{
+                            Manifest.permission.RECORD_AUDIO,
+                            Manifest.permission.MODIFY_AUDIO_SETTINGS,
+                            Manifest.permission.WRITE_EXTERNAL_STORAGE },
+                    AUDIO_ECHO_REQUEST);
+        }
+    }
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
+                                           @NonNull int[] grantResults) {
+
+        /*
+         * if any permission failed, the sample could not play
+         */
+        if (AUDIO_ECHO_REQUEST != 0) {
```

Comment:
wouldn't this be always true? (because private static final int AUDIO_ECHO_REQUEST = 0;)